File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        def backtrack(row, col, path):
            if len(path) == len(word):
                return True

            res = False
            if (0 <= row < m
                and 0 <= col < n
                and (row, col) not in path
                and board[row][col] == word[len(path)]
            ):
                path.append((row, col))
                for y, x in directions:
                    if backtrack(row + y, col + x, path):
                        return True
                path.pop()

            return False
            
        for row in range(m):
            for col in range(n):
                if backtrack(row, col, []):
                    return True

        return False

# Time complexity: O(m * n * 3^l) where l is the length of the word to be matched
# Space complexity: O(l) for the recursion stack

File: 0079-word-search/0079-word-search.java
class Solution {
    private char[][] board;
    private int ROWS;
    private int COLS;
    private int[][] directions;

    public boolean exist(char[][] board, String word) {
        this.board = board;
        this.ROWS = board.length;
        this.COLS = board[0].length;
        this.directions = new int[][] { {0, 1}, {0, -1}, {1, 0}, {-1, 0} };

        for (int row = 0; row < this.ROWS; row++) {
            for (int col = 0; col < this.COLS; col++) {
                if (this.backtrack(row, col, word, 0)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean backtrack(int row, int col, String word, int index) {
        if (index >= word.length()) {
            return true;
        }

        if (
            row < 0 ||
            row == this.ROWS ||
            col < 0 ||
            col == this.COLS ||
            this.board[row][col] != word.charAt(index)
        ) {
            return false;
        }

        boolean res = false;
        this.board[row][col] = '#';

        for (int[] d : this.directions) {
            if (this.backtrack(
                row + d[0],
                col + d[1],
                word, index + 1
            )) {
                res = true;
                break;
            }
        }

        this.board[row][col] = word.charAt(index);
        return res;
    }
}